# 📖 스프링과 문제 해결 - 예외 처리, 반복
<br/><br/>
# 🔎 체크 예외와 인터페이스
서비스 계층은 가급적 특정 구현 기술에 의존하지 않고, 순수하게 유지하는 것이 좋다. 이렇게 하려면 예외에 대한 의존도 함께 해결해야 한다.<br/><br/>
서비스가 처리할 수 없으므로 리포지토리가 던지는 `SQLException` 체크 예외를 런타임 예외로 전환해서 서비스 계층에 던지면 된다.<br/>
이렇게 하면 서비스 계층이 해당 예외를 무시할 수 있기 때문에, 특정 구현 기술에 의존하는 부분을 제거하고 서비스 계층을 순수하게 유지할 수 있다.<br/><br/>
### 인터페이스 도입
먼저 `MemberRepository` 인터페이스를 도입해서 구현 기술을 쉽게 변경할 수 있게 한다.<br/><br/>
![image](https://github.com/jang9205/spring_study/assets/123082616/8c1f213d-8933-424c-828c-e82ecc2b285f)<br/>
- 인터페이스를 도입하면 `MemberService`는 `MemberRepository` 인터페이스에만 의존하면 된다.
- 이제 구현 기술을 변경하고 싶으면 DI를 사용해서 `MemberService` 코드의 변경 없이 구현 기술을 변경할 수 있다.
<br/><br/>
#### 체크 예외와 인터페이스
기존에는 `SQLException`이 체크 예외이기 때문에 인터페이스를 만들지 않았다. 인터페이스의 구현체가 체크 예외를 던지려면, 인터페이스 메서드에 먼저 체크 예외를 던지는 부분이 선언되어 있어야 한다. 그래야 구현 클래스의 메서드도 체크 예외를 던질 수 있다.<br/><br/>
구현 기술을 쉽게 변경하기 위해서 인터페이스를 도입하더라도 `SQLException`과 같은 특정 구현 기술에 종속적인 체크 예외를 사용하게 되면 인터페이스에도 해당 예외를 포함해야 한다.<br/>
하지만 이것은 순수한 인터페이스가 아니다. JDBC 기술에 종속적인 인터페이스일 뿐이다. 인터페이스를 만드는 목적은 구현체를 쉽게 변경하기 위함인데, 이미 인터페이스가 특정 구현 기술에 오염이 되어 버렸다. 향후 JDBC가 아닌 다른 기술로 변경한다면 인터페이스 자체를 변경해야 한다.<br/><br/>
#### 런타임 예외와 인터페이스
런타임 예외는 이런 부분에서 자유롭다. 인터페이스에 런타임 예외를 따로 선언하지 않아도 된다. 따라서 인터페이스가 특정 기술에 종속적일 필요가 없다.<br/><br/>
# 🔎 런타임 예외 적용
`java/hello/jdbc/repository/MemberRepository`
```java
public interface MemberRepository {
    Member save(Member member);
    Member findById(String memberId);
    void update(String memberId, int money);
    void delete(String memberId);
}
```
<br/><br/>
`java/hello/jdbc/repository/ex/MyDbException`
```java
public class MyDbException extends RuntimeException{
    public MyDbException() {
    }

    public MyDbException(String message) {
        super(message);
    }

    public MyDbException(String message, Throwable cause) {
        super(message, cause);
    }

    public MyDbException(Throwable cause) {
        super(cause);
    }
}
```
`RuntimeException`을 상속받았다. 따라서 `MyDbException`은 런타임(언체크) 예외가 된다.<br/><br/>
`java/hello/jdbc/repository/MemberRepositoryV4_1`
```java
@Slf4j
public class MemberRepositoryV4_1 implements MemberRepository {

    ...

    @Override
    public Member save(Member member) {

        ...

        } catch (SQLException e) {
            throw new MyDbException(e);

        ...

    @Override
    public Member findById(String memberId) {

        ...

        } catch (SQLException e) {
            throw new MyDbException(e);

        ...
```
`MemberRepository` 인터페이스를 구현한다.<br/>
또한 `SQLException` 체크 예외를 `MyDbException` 런타임 예외로 변환해서 던진다.<br/><br/>
기존 예외를 생성자를 통해서 포함하고 있는 것을 확인할 수 있다. 그래야 예외를 출력했을 때 원인이 되는 기존 예외도 함께 확인할 수 있다.<br/>
예외를 출력했을 때 스택 트레이스를 통해 둘 다 확인할 수 있다.<br/><br/>
`java/hello/jdbc/service/MemberServiceV4`
```java
@Slf4j
public class MemberServiceV4 {

    private final MemberRepository memberRepository;

    public MemberServiceV4(MemberRepository memberRepository) {
        this.memberRepository = memberRepository;
    }

    @Transactional
    public void accountTransfer(String fromId, String toId, int money) {
        bizLogic(fromId, toId, money);
    }

    private void bizLogic(String fromId, String toId, int money) {

        ...
```
`MemberRepository` 인터페이스에 의존하도록 코드를 변경했다.<br/>
또한, 메서드에서 `throws SQLException` 부분이 제거된 것을 확인할 수 있다. 드디어 순수한 서비스를 완성했다.<br/><br/>
`test/java/hello/jdbc/service/MemberServiceV4Test`
```java
    ...

    @Autowired
    private MemberRepository memberRepository;
    @Autowired
    private MemberServiceV4 memberService;

    @TestConfiguration
    static class TestConfig {

        private final DataSource dataSource;

        public TestConfig(DataSource dataSource) {
            this.dataSource = dataSource;
        }

        @Bean
        MemberRepository memberRepository() {
            return new MemberRepositoryV4_1(dataSource);
        }

        @Bean
        MemberServiceV4 memberServiceV4() {
            return new MemberServiceV4(memberRepository());
        }
    }

    ...

    @Test
    @DisplayName("정상 이체")
    void accountTransfer() {

    ...

    @Test
    @DisplayName("이체중 예외 발생")
    void accountTransferEx() {

    ...
```
<br/><br/>
# 🔎 데이터 접근 예외 직접 만들기
데이터베이스 오류에 따라서 특정 예외는 복구하고 싶을 수 있다.<br/>
예를 들어서 회원 가입시 DB에 이미 같은 ID가 있으면 ID 뒤에 숫자를 붙여서 새로운 ID를 만들어야 한다고 가정해보자.<br/>
ID를 `hello`라고 가입 시도 했는데, 이미 같은 아이디가 있으면 `hello12345`와 같이 뒤에 임의의 숫자를 붙여서 가입하는 것이다.<br/><br/>
데이터를 DB에 저장할 때 같은 ID가 이미 데이터베이스에 저장되어 있다면, 데이터베이스는 오류 코드를 반환하고, 이 오류 코드를 받은 JDBC 드라이버는 `SQLException`을 던진다.<br/>
그리고 `SQLException`에는 데이터베이스가 제공하는 `errorCode`라는 것이 들어있다.<br/><br/>
#### 데이터베이스 오류 코드 그림
![image](https://github.com/jang9205/spring_study/assets/123082616/9cc157bd-4274-49bb-b3a5-53561a9f6fe4)<br/><br/>
#### H2 데이터베이스의 키 중복 오류 코드
```java
e.getErrorCode() == 23505
```
`SQLException` 내부에 들어있는 `errorCode`를 활용하면 데이터베이스에서 어떤 문제가 발생했는지 확인할 수 있다.<br/><br/>
#### 예) 키 중복 오류 코드
- H2 DB: `23505`
- MySQL: `1062`
<br/><br/>같은 오류여도 각각의 데이터베이스마다 정의된 오류 코드가 다르다. 따라서 오류 코드를 사용할 때는 데이터베이스 메뉴얼을 확인해야 한다.<br/><br/>
#### H2 데이터베이스 오류 코드
https://www.h2database.com/javadoc/org/h2/api/ErrorCode.html<br/><br/>
서비스 계층에서는 예외 복구를 위해 키 중복 오류를 확인할 수 있어야 한다. 그래야 새로운 ID를 만들어서 다시 저장을 시도할 수 있기 때문이다. 이러한 과정이 바로 예외를 확인해서 복구하는 과정이다. 리포지토리는 `SQLException`을 서비스 계층에 던지고 서비스 계층은 이 예외의 오류 코드를 확인해서 키 중복 오류(`23505`)인 경우 새로운 ID를 만들어서 다시 저장하면 된다.<br/><br/>
그런데 `SQLException`에 들어있는 오류 코드를 활용하기 위해 `SQLException`을 서비스 계층으로 던지게 되면, 서비스 계층이 `SQLException`이라는 JDBC 기술에 의존하게 되면서, 지금까지 우리가 고민했던 서비스 계층의 순수성이 무너진다.<br/><br/>
이 문제를 해결하려면 앞서 배운 것처럼 리포지토리에서 예외를 변환해서 던지면 된다.<br/>
`SQLException` -> `MyDuplicateKeyException`<br/><br/>
`java/hello/jdbc/repository/ex/MyDuplicateKeyException`
```java
public class MyDuplicateKeyException extends MyDbException {

    public MyDuplicateKeyException() {
    }

    public MyDuplicateKeyException(String message) {
        super(message);
    }

    public MyDuplicateKeyException(String message, Throwable cause) {
        super(message, cause);
    }

    public MyDuplicateKeyException(Throwable cause) {
        super(cause);
    }
}
```
- 기존에 사용했던 `MyDbException`을 상속받아서 의미있는 계층을 형성한다. 이렇게하면 데이터베이스 관련 예외라는 계층을 만들 수 있다.
- `MyDuplicateKeyException`이라는 이름을 지었다. 이 예외는 데이터 중복의 경우에만 던져야 한다.
- 이 예외는 우리가 직접 만든 것이기 때문에, JDBC나 JPA 같은 특정 기술에 종속적이지 않다. 따라서 이 예외를 사용하더라도 서비스 계층의 순수성을 유지할 수 있다.
#### 
실제 예제 코드를 만들어서 확인해보자.<br/><br/>
`test/java/hello/jdbc/exception/translator/ExTranslatorV1Test`
```java
@Slf4j
public class ExTranslatorV1Test {

    Repository repository;
    Service service;

    @BeforeEach
    void init() {
        DriverManagerDataSource dataSource = new DriverManagerDataSource(URL, USERNAME, PASSWORD);
        repository = new Repository(dataSource);
        service = new Service(repository);
    }

    @Test
    void duplicateKeySave() {
        service.create("myId");
        service.create("myId"); //같은 ID 저장 시도
    }
    @Slf4j
    @RequiredArgsConstructor
    static class Service {
        private final Repository repository;

        public void create(String memberId) {
            try {
                repository.save(new Member(memberId, 0));
                log.info("saveId={}", memberId);
            } catch (MyDuplicateKeyException e) {
                log.info("키 중복, 복구 시도");
                String retryId = generateNewId(memberId);
                log.info("retryId={}", retryId);
                repository.save(new Member(retryId, 0));
            } catch (MyDbException e) {
                log.info("데이터 접근 계층 예외", e);
                throw e;
            }
        }

        private String generateNewId(String memberId) {
            return memberId + new Random().nextInt(10000);
        }
    }

    @RequiredArgsConstructor
    static class Repository {
        private final DataSource dataSource;

        public Member save(Member member) {
            String sql = "insert into member(member_id, money) values(?,?)";
            Connection con = null;
            PreparedStatement pstmt = null;

            try {
                con = dataSource.getConnection();
                pstmt = con.prepareStatement(sql);
                pstmt.setString(1, member.getMemberId());
                pstmt.setInt(2, member.getMoney());
                pstmt.executeUpdate();
                return member;
            } catch (SQLException e) {
                //h2 db
                if (e.getErrorCode() == 23505) {
                    throw new MyDuplicateKeyException(e);
                }
                throw new MyDbException(e);
            } finally {
                JdbcUtils.closeStatement(pstmt);
                JdbcUtils.closeConnection(con);
            }
        }
    }
}
```
**리포지토리**<br/>
`e.getErrorCode() == 23505`: 오류 코드가 키 중복 오류(`23505`)인 경우 `MyDuplicateKeyException`을 새로 만들어서 서비스 계층에 던진다.<br/>
나머지 경우 기존에 만들었던 `MyDbException`을 던진다.<br/><br/>
**서비스**<br/>
처음에 저장을 시도한다. 만약 리포지토리에서 `MyDuplicateKeyException` 예외가 올라오면 이 예외를 잡는다. 예외를 잡아서 `generateNewId(memberId)`로 새로운 ID 생성을 시도한다.<br/>
만약 복구할 수 없는 예외(`MyDbException`)면 로그만 남기고 다시 예외를 던진다.<br/><br/>
실행해보면 다음과 같은 로그를 확인할 수 있다. 
```
Service - saveId=myId
Service - 키 중복, 복구 시도
Service - retryId=myId492
```
같은 ID를 저장했지만, 중간에 예외를 잡아서 복구한 것을 확인할 수 있다.<br/><br/>
#### 정리
- SQL ErrorCode로 데이터베이스에 어떤 오류가 있는지 확인할 수 있었다.
- 예외 변환을 통해 `SQLException`을 특정 기술에 의존하지 않는 직접 만든 예외인 `MyDuplicateKeyException`로 변환 할 수 있었다.
- 리포지토리 계층이 예외를 변환해준 덕분에 서비스 계층은 특정 기술에 의존하지 않는 `MyDuplicateKeyException`을 사용해서 문제를 복구하고, 서비스 계층의 순수성도 유지할 수 있었다.
<br/><br/>
#### 남은 문제
- SQL ErrorCode는 각각의 데이터베이스 마다 다르다. 결과적으로 데이터베이스가 변경될 때 마다 ErrorCode도 모두 변경해야 한다.
- 데이터베이스가 전달하는 오류는 키 중복 뿐만 아니라 락이 걸린 경우, SQL 문법에 오류 있는 경우 등 수십 수백가지 오류 코드가 있다. 이 모든 상황에 맞는 예외를 지금처럼 다 만들어야 할까?
<br/><br/>
# 🔎 스프링 예외 추상화 이해
스프링은 앞서 설명한 문제들을 해결하기 위해 데이터 접근과 관련된 예외를 추상화해서 제공한다.<br/><br/>
#### 스프링 데이터 접근 예외 계층
![image](https://github.com/jang9205/spring_study/assets/123082616/bdab1ed2-0d09-486b-a191-0bf447506e74)<br/>
- 스프링은 데이터 접근 계층에 대한 수십 가지 예외를 정리해서 일관된 예외 계층을 제공한다.
- 각각의 예외는 특정 기술에 종속적이지 않게 설계되어 있다. 따라서 서비스 계층에서도 스프링이 제공하는 예외를 사용하면 된다.
- JDBC나 JPA를 사용할 때 발생하는 예외를 스프링이 제공하는 예외로 변환해주는 역할도 스프링이 제공한다.<br/><br/>
- 예외의 최고 상위는 `org.springframework.dao.DataAccessException`이다. 런타임 예외를 상속 받았기 때문에 스프링이 제공하는 데이터 접근 계층의 모든 예외는 런타임 예외이다.
- `DataAccessException`은 크게 `NonTransient` 예외와 `Transient` 예외로 구분이 가능하다.
  - `Transient`는 일시적이라는 뜻이다. `Transient` 하위 예외는 동일한 SQL을 다시 시도했을 때 성공할 가능성이 있다.
    - 예를 들어서 쿼리 타임아웃, 락과 관련된 오류들이다. 이런 오류들은 데이터베이스 상태가 좋아지거나, 락이 풀렸을 때 다시 시도하면 성공할 수도 있다.
  - `NonTransient`는 일시적이지 않다는 뜻이다. 같은 SQL을 그대로 반복해서 실행하면 실패한다.
    - SQL 문법 오류, 데이터베이스 제약조건 위배 등이 있다.
<br/><br/>스프링 메뉴얼에 모든 예외가 정리되어 있지는 않기 때문에 코드를 직접 열어서 확인해보는 것이 필요하다.<br/><br/>
### 스프링이 제공하는 예외 변환기
스프링은 데이터베이스에서 발생하는 오류 코드를 스프링이 정의한 예외로 자동으로 변환해주는 변환기를 제공한다.<br/><br/>
`test/java/hello/jdbc/exception/translator/SpringExceptionTranslatorTest`
```java
@Slf4j
public class SpringExceptionTranslatorTest {

    DataSource dataSource;

    @BeforeEach
    void init() {
        dataSource = new DriverManagerDataSource(URL, USERNAME, PASSWORD);
    }

    @Test
    void exceptionTranslator() {

        String sql = "select bad grammar";

        try {
            Connection con = dataSource.getConnection();
            PreparedStatement stmt = con.prepareStatement(sql);
            stmt.executeQuery();
        } catch (SQLException e) {
            assertThat(e.getErrorCode()).isEqualTo(42122);

            SQLErrorCodeSQLExceptionTranslator exTranslator = new SQLErrorCodeSQLExceptionTranslator(dataSource);

            //BadSqlGrammarException
            DataAccessException resultEx = exTranslator.translate("select", sql, e);
            log.info("resultEx", resultEx);
            assertThat(resultEx.getClass()).isEqualTo(BadSqlGrammarException.class);
        }
    }
}
```
- `translate()` 메서드의 첫번째 파라미터는 읽을 수 있는 설명이고, 두번째는 실행한 sql, 마지막은 발생된 `SQLException`을 전달하면 된다. 이렇게 하면 적절한 스프링 데이터 접근 계층의 예외로 변환해서 반환해준다.
- 예제에서는 SQL 문법이 잘못되었으므로 `BadSqlGrammarException`을 반환하는 것을 확인할 수 있다.
- 눈에 보이는 반환 타입은 최상위 타입인 `DataAccessException`이지만 실제로는 `BadSqlGrammarException` 예외가 반환된다.
#### 
<br/><br/>각각의 DB마다 SQL ErrorCode는 다르다. 그런데 스프링은 어떻게 각각의 DB가 제공하는 SQL ErrorCode까지 고려해서 예외를 변환할 수 있을까?<br/><br/>
비밀은 바로 다음 파일에 있다.<br/>
`sql-error-codes.xml`
```xml
	<bean id="H2" class="org.springframework.jdbc.support.SQLErrorCodes">
		<property name="badSqlGrammarCodes">
			<value>42000,42001,42101,42102,42111,42112,42121,42122,42132</value>
		</property>
		<property name="duplicateKeyCodes">
			<value>23001,23505</value>
		</property>
		<property name="dataIntegrityViolationCodes">
			<value>22001,22003,22012,22018,22025,23000,23002,23003,23502,23503,23506,23507,23513</value>
		</property>
		<property name="dataAccessResourceFailureCodes">
			<value>90046,90100,90117,90121,90126</value>
		</property>
		<property name="cannotAcquireLockCodes">
			<value>50200</value>
		</property>
	</bean>
```
스프링 SQL 예외 변환기는 SQL ErrorCode를 이 파일에 대입해서 어떤 스프링 데이터 접근 예외로 전환해야 할지 찾아낸다. 예를 들어서 H2 데이터베이스에서 `42000`이 발생하면 `badSqlGrammarCodes`이기 때문에`BadSqlGrammarException`을 반환한다.<br/><br/>
#### 정리
- 스프링은 데이터 접근 계층에 대한 일관된 예외 추상화를 제공한다.
- 스프링은 예외 변환기를 통해서 `SQLException`의 `ErrorCode`에 맞는 적절한 스프링 데이터 접근 예외로 변환해준다.
- 만약 서비스, 컨트롤러 계층에서 예외 처리가 필요하면 특정 기술에 종속적인 `SQLException` 같은 예외를 직접 사용하는 것이 아니라, 스프링이 제공하는 데이터 접근 예외를 사용하면 된다.
- 스프링 예외 추상화 덕분에 특정 기술에 종속적이지 않게 되었다.
<br/><br/>
# 🔎 스프링 예외 추상화 적용
이제 애플리케이션에 스프링이 제공하는 데이터 접근 예외 추상화와 SQL 예외 변환기를 적용해보자.<br/><br/>
`java/hello/jdbc/repository/MemberRepositoryV4_2`
```java
@Slf4j
public class MemberRepositoryV4_2 implements MemberRepository {

    private final DataSource dataSource;
    private final SQLExceptionTranslator exTranslator;

    public MemberRepositoryV4_2(DataSource dataSource) {
        this.dataSource = dataSource;
        this.exTranslator = new SQLErrorCodeSQLExceptionTranslator(dataSource);
    }

    @Override
    public Member save(Member member) {

        ...

        } catch (SQLException e) {
            throw exTranslator.translate("save", sql, e);

        ...

    @Override
    public Member findById(String memberId) {

        ...

        } catch (SQLException e) {
            throw exTranslator.translate("findById", sql, e);

        ...
```
기존 코드에서 스프링 예외 변환기를 사용하도록 변경되었다.<br/><br/>
스프링이 예외를 추상화해준 덕분에, 서비스 계층은 특정 리포지토리의 구현 기술과 예외에 종속적이지 않게 되었다.<br/>
추가로 서비스 계층에서 예외를 잡아서 복구해야 하는 경우, 예외가 스프링이 제공하는 데이터 접근 예외로 변경되어서 서비스 계층에 넘어오기 때문에 필요한 경우 예외를 잡아서 복구하면 된다.<br/><br/>
# 🔎 JDBC 반복 문제 해결 - JdbcTemplate
지금까지 서비스 계층의 순수함을 유지하기 위해 수 많은 노력을 했고, 덕분에 서비스 계층의 순수함을 유지하게 되었다. 이번에는 리포지토리에서 JDBC를 사용하기 때문에 발생하는 반복 문제를 해결해보자.<br/><br/>
**JDBC 반복 문제**
- 커넥션 조회, 커넥션 동기화
- PreparedStatement 생성 및 파라미터 바인딩
- 쿼리 실행
- 결과 바인딩
- 예외 발생 시 스프링 예외 변환기 실행
- 리소스 종료
#### 
리포지토리의 각각의 메서드를 살펴보면 상당히 많은 부분이 반복된다. 이런 반복을 효과적으로 처리하는 방법이 바로 템플릿 콜백 패턴이다.<br/>
스프링은 JDBC의 반복 문제를 해결하기 위해 `JdbcTemplate`이라는 템플릿을 제공한다.<br/><br/>
`java/hello/jdbc/repository/MemberRepositoryV5`
```java
@Slf4j
public class MemberRepositoryV5 implements MemberRepository {

    private final JdbcTemplate template;

    public MemberRepositoryV5(DataSource dataSource) {
        this.template = new JdbcTemplate(dataSource);
    }

    @Override
    public Member save(Member member) {
        String sql = "insert into member(member_id, money) values (?, ?)";
        template.update(sql, member.getMemberId(), member.getMoney());
        return member;
    }

    @Override
    public Member findById(String memberId) {
        String sql = "select * from member where member_Id = ?";
        return template.queryForObject(sql, memberRowMapper(), memberId);
    }

    @Override
    public void update(String memberId, int money) {
        String sql = "update member set money=? where member_id=?";
        template.update(sql, money, memberId);
    }

    @Override
    public void delete(String memberId) {
        String sql = "delete from member where member_id=?";
        template.update(sql, memberId);
    }

    private RowMapper<Member> memberRowMapper() {
        return (rs, rowNum) -> {
            Member member = new Member();
            member.setMemberId(rs.getString("member_id"));
            member.setMoney(rs.getInt("money"));
            return member;
        };
    }
}
```
`JdbcTemplate`은 JDBC로 개발할 때 발생하는 반복을 대부분 해결해준다.<br/>
그 뿐만 아니라 지금까지 학습했던, **트랜잭션을 위한 커넥션 동기화**는 물론이고, 예외 발생 시 **스프링 예외 변환기**도 자동으로 실행해준다.